x-pack/plugins/canvas/public/components/es_fields_select/es_fields_select.tsx

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useState, useEffect } from 'react';
+import { isEqual } from 'lodash';
+import { getFields } from '../../lib/es_service';
+import {
+  ESFieldsSelect as Component,
+  ESFieldsSelectProps as Props,
+} from './es_fields_select.component';
+
+type ESFieldsSelectProps = Omit<Props, 'fields'> & { index: string };
+
+export const ESFieldsSelect: React.FunctionComponent<ESFieldsSelectProps> = (props) => {
+  const { index, selected, onChange } = props;
+  const [fields, setFields] = useState<string[]>([]);
+  useEffect(() => {
+    getFields(index).then((newFields) => {
+      if (!isEqual(newFields, fields)) {
+        setFields(newFields || []);
+        const filteredSelected = selected.filter((option) => (newFields || []).includes(option));
```

Comment:
I'm no sure why we need to do this filtering on the `getFields` result.  And the hook being dependent on the `selected` state makes it fire off the `getFields` request every time a new value is selected.

Reply:
@crob611, I've updated the code in a way it will not fire `getFields` if the index is equal to the previous index. 
What about `selected`, we need to check if some options are present in new fields.  I've added the check of equality. If arrays of previously selected options and those, which came are the same, it is not updating an array.

Could you, please, review the current PR again. Thanks)